src/compiler/checker.ts

```diff
@@ -19048,24 +19048,37 @@ namespace ts {
             else if (args) {
                 let min = Number.POSITIVE_INFINITY;
                 let max = Number.NEGATIVE_INFINITY;
+                let belowArgCount = Number.NEGATIVE_INFINITY;
+                let aboveArgCount = Number.POSITIVE_INFINITY;
+
+                let argCount = args.length;
                 for (const sig of signatures) {
-                    min = Math.min(min, getMinArgumentCount(sig));
-                    max = Math.max(max, getParameterCount(sig));
+                    const minCount = getMinArgumentCount(sig);
+                    const maxCount = getParameterCount(sig);
+                    if (minCount < argCount && minCount > belowArgCount) belowArgCount = minCount;
+                    if (argCount < maxCount && maxCount < aboveArgCount) aboveArgCount = maxCount;
+                    min = Math.min(min, minCount);
+                    max = Math.max(max, maxCount);
                 }
+
                 const hasRestParameter = some(signatures, hasEffectiveRestParameter);
-                const hasSpreadArgument = getSpreadArgumentIndex(args) > -1;
-                const paramCount = hasRestParameter ? min :
+                const paramRange = hasRestParameter ? min :
                     min < max ? min + "-" + max :
                     min;
-                let argCount = args.length;
-                if (argCount <= max && hasSpreadArgument) {
-                    argCount--;
-                }
-                const error = hasRestParameter && hasSpreadArgument ? Diagnostics.Expected_at_least_0_arguments_but_got_1_or_more :
-                    hasRestParameter ? Diagnostics.Expected_at_least_0_arguments_but_got_1 :
-                    hasSpreadArgument ? Diagnostics.Expected_0_arguments_but_got_1_or_more :
-                    Diagnostics.Expected_0_arguments_but_got_1;
-                diagnostics.add(createDiagnosticForNode(node, error, paramCount, argCount));
+                const hasSpreadArgument = getSpreadArgumentIndex(args) > -1;
+                if (argCount <= max && hasSpreadArgument) argCount--;
+
+                if (hasRestParameter || hasSpreadArgument) {
+                    const error = hasRestParameter && hasSpreadArgument ? Diagnostics.Expected_at_least_0_arguments_but_got_1_or_more :
+                        hasRestParameter ? Diagnostics.Expected_at_least_0_arguments_but_got_1 : Diagnostics.Expected_0_arguments_but_got_1_or_more;

```

Comment:
Please revert this line-join and put `Diagnostics.Expected 0 arguments but got 1 or more` back on its own line.